### PR TITLE
Candidate patches

### PR DESCRIPTION
**Pull request by titu1994**
# Bugfixes
- Correctly call `multi_test_epoch_end()` when performing `trainer.test()`.
- Raise warning per call instead of raising NotImplementedError() when `multi_*` methods are not implemented but subclass also does not override `validation_epoch_end()` and `test_epoch_end()`. This is an uncertain state between unknowingly supporting multi data loaders but not overriding any of the required methods.
- Add min_lr to all NeMo schedulers so that min_lr is properly preserved for all schedulers.
- Patch schedulers to properly support target / cls instead of `name`. 
- Schedulers now fail with warning that they were not initialized instead of raising ValueError. This is necessary when re-instantiating models from configs but data loaders are not available to compute effective `max_steps`.
- Optimizers now preserve the config passed to them via deepcopy.

# Improvements
- Add helper method to correctly test a model on the test set.
- Config check when optimizer parameters are passed via their Param classes.
- Add full test suite for Optimizers and Schedulers

### CODE REVIEW

File: nemo/core/classes/modelPT.py [hunk 1]
```diff
@@ -583,19 +583,79 @@ def test_epoch_end(
...
+            "`test_epoch_end(outputs)."
+        )
 
     def get_validation_dataloader_prefix(self, dataloader_idx=0):
         return self._validation_names[dataloader_idx]
 
     def get_test_dataloader_prefix(self, dataloader_idx=0):
         return self._test_names[dataloader_idx]
 
+    def prepare_test(self) -> bool:
```
Comment (okuchaiev) [inline nemo/core/classes/modelPT.py:611]: is this patch around PTL?

Comment (titu1994) [reply nemo/core/classes/modelPT.py:611]: The optimization fix is over AMP (I'm looking for a better way to solve this), and the DDP one is PTL

File: nemo/core/config/optimizers.py [hunk 2]
```diff
@@ -238,6 +240,11 @@ def get_optimizer_config(name: str, **kwargs: Optional[Dict[str, Any]]) -> Optim
         )
 
     scheduler_params = AVAILABLE_OPTIMIZER_PARAMS[name]
+
+    if kwargs is not None and len(kwargs) != 0:
+        kwargs = OmegaConf.create(kwargs)
+        OmegaConf.merge(scheduler_params(), kwargs)
```
Comment (blisc) [inline nemo/core/config/optimizers.py:246]: I've been thinking more about this construction. We honestly could do away with dataclasses if we just use `hydra.util.instantiate` instead and pass normal python arugments as opposed to DictConfigs to our classes.

File: nemo/core/config/optimizers.py [hunk 3]
```diff
@@ -238,6 +240,11 @@ def get_optimizer_config(name: str, **kwargs: Optional[Dict[str, Any]]) -> Optim
...
+
     scheduler_params = partial(scheduler_params, **kwargs)
```
Comment (blisc) [inline nemo/core/config/optimizers.py:248]: You can avoid the partial if you want
```python
    scheduler_params = OmegaConf.structured(AVAILABLE_OPTIMIZER_PARAMS[name])
    if kwargs is not None and len(kwargs) != 0:
        kwargs = OmegaConf.create(kwargs)
        scheduler_params = OmegaConf.merge(scheduler_params, kwargs)
    return scheduler_params  # A DictConfig not a (class pointer/definition, not sure what to call this)
```

File: nemo/core/config/optimizers.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/config/optimizers.py:246]: Instantiate requires the assumption you are following the `target` and `params` pattern which causes very deep nest when trying to override parameters for optimizers - `model.optim.sched.params.weight_decay` which is untenable.

File: nemo/core/config/optimizers.py (hunk 3 above)
Comment (titu1994) [reply nemo/core/config/optimizers.py:248]: Scheduler parameter dataclasses are somewhat special, they dont hold the `max_steps` kwarg even though their constructor does have this value (in most but not all constructors). I have to defer its calculation until we resolve either `iters_per_batch` or the actual passed `max_steps` and finally instantiate the constructor of the scheduler with the union of the params and this `max_steps`.

**Comment by titu1994** [comment]
@blisc @okuchaiev I've simplified optimizer construction, so no more need to explicitly call `setup_optimization()` inside any constructor. This has the additional benefit that we dont need to have edge cases for AMP during testing.

Also added a patch for MFCC after recent tts PR which altered signature of `get_seq_length()` to account for MFCC signature. (It doesnt get caught in MatchboxNet tests since torchaudio is not installed on CI so I force it to switch to MelSpec).

File: nemo/core/config/optimizers.py (hunk 2 above)
Comment (blisc) [reply nemo/core/config/optimizers.py:246]: This exposes an issue with the construction of our yaml files, not a problem with hydra.
Nothing is stopping us from exposing it at `sched.params.weight_decay`, or `sched.weight_decay`, or even `weight_decay`.

Comment (titu1994) [reply nemo/core/config/optimizers.py:246]: Well we need to make certain assumptions regarding this, and using classpath everywhere is problematic in all cases other than pure productions configs (which we won't even need there since we will simply be using config objects) so for research flexibility the current schema is practical and useful enough.

File: nemo/core/optim/lr_scheduler.py [hunk 4]
```diff
@@ -434,17 +430,32 @@ def prepare_lr_scheduler(
         logging.info('Scheduler not initialized as no `sched` config supplied to setup_optimizer()')
         return None
 
-    # Get name of the scheduler
-    scheduler_name = scheduler_config['name']
-
     # Try instantiation of scheduler params from config class path
     try:
-        scheduler_conf = hydra.utils.instantiate(scheduler_args)
```
Comment (omry) [inline nemo/core/optim/lr_scheduler.py:442]: You replace one line of instantiate with many lines which are more complicated.
I am curious as to the motivation here.
